### PR TITLE
Ruby: also treat included/prepended modules as subclasses

### PR DESCRIPTION
**Pull request by aibaars**
When resolving a call we should also consider any included/prepended modules as ancestors

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll [hunk 1]
```diff
@@ -197,7 +197,7 @@ private predicate instanceMethodCall(CfgNodes::ExprNodes::CallCfgNode call, Modu
     or
     // When we don't know the exact type, it could be any sub class
     exact = false and
-    tp.getSuperClass+() = m
+    [tp.getSuperClass+(), tp.getAnIncludedModule+(), tp.getAPrependedModule+()] = m
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:200]: I guess it should really be
```
tp.getACludedModule+()
```
where `getACludedModule` is the union of `getAnIncludedModule` and `getAPrependedModule`?

File: ruby/ql/lib/codeql/ruby/frameworks/http_clients/Faraday.qll [hunk 2]
```diff
@@ -37,7 +37,8 @@ class FaradayHttpRequest extends Http::Client::Request::Range, DataFlow::CallNod
         API::getTopLevelMember("Faraday").getInstance()
       ] and
     requestNode =
-      connectionNode.getReturn(["get", "head", "delete", "post", "put", "patch", "trace"]) and
+      connectionNode
+          .getReturn(["get", "head", "delete", "post", "put", "patch", "trace", "run_request"]) and
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/frameworks/http_clients/Faraday.qll:41]: Perhaps this change belongs to another PR?

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/frameworks/http_clients/Faraday.qll:41]: I could have made a separate PR for it indeed. It was such a minor change that I left it in .

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll (hunk 1 above)
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:200]: Perhaps even `getASuperCludedModule+` , a perhaps better named `getAnAncestor+`()?

File: ruby/ql/lib/codeql/ruby/ast/Module.qll [hunk 3]
```diff
@@ -23,6 +23,14 @@ class Module extends TModule {
   /** Gets an `include`d module. */
   Module getAnIncludedModule() { result = getAnIncludedModule(this) }
 
+  /** Gets a direct ancestor of this module. */
+  Module getADirectAncestor() {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ast/Module.qll:27]: I think we tend to use "immediate" instead of "direct"

File: ruby/ql/lib/codeql/ruby/ast/Module.qll [hunk 4]
```diff
@@ -23,6 +23,14 @@ class Module extends TModule {
...
+    result = [getAnIncludedModule(this), getAPrependedModule(this), getSuperClass(this)]
+  }
+
+  /** Gets an ancestor of this module. */
+  Module getAnAncestor() { result = this.getADirectAncestor+() }
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ast/Module.qll:32]: I would `pragma[inline]` this.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll [hunk 5]
```diff
@@ -464,9 +461,9 @@ private module Cached {
       )
     )
     or
-    exists(Module superClass, string method |
-      superCall(call, superClass, method) and
-      result = lookupMethod(superClass, method)
+    exists(Module cls, string method |
+      superCall(call, cls, method) and
+      result = lookupMethod(cls.getAnImmediateAncestor+(), method)
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:466]: Shouldn't this just be `getAnImmediateAncestor()` without the `+`? `lookupMethod` looks up in ancestors already.

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:466]: Good point!

**Review by asgerf — APPROVED** [review]
